### PR TITLE
Fix #1455 NPE when retrieving BibteX data from DOI

### LINKED ISSUE #1455: DOI Fetcher throws NPE if no result was found
JabRef dev version on Windows 10

Steps to reproduce:

1. Entry with DOI for which the fetcher can't find bibtex data, for example: 10.2307/2372901
2. Click "Get BibTeX data from DOI"
3. Results in no warning but an NPE:
````
Exception in thread "AWT-EventQueue-0" java.lang.NullPointerException
	at java.util.Objects.requireNonNull(Objects.java:203)
	at net.sf.jabref.importer.fetcher.DOItoBibTeXFetcher.getEntryFromDOI(DOItoBibTeXFetcher.java:90)
	at net.sf.jabref.gui.mergeentries.MergeEntryDOIDialog.<init>(MergeEntryDOIDialog.java:75)
	at net.sf.jabref.gui.entryeditor.FieldExtraComponents.lambda$getDoiExtraComponent$339(FieldExtraComponents.java:199)
	at javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:2022)
	at javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2348)
	at javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:402)
	at javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:259)
	at javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:252)
	at java.awt.Component.processMouseEvent(Component.java:6535)
	at javax.swing.JComponent.processMouseEvent(JComponent.java:3324)
	at java.awt.Component.processEvent(Component.java:6300)
	at java.awt.Container.processEvent(Container.java:2236)
	at java.awt.Component.dispatchEventImpl(Component.java:4891)
	at java.awt.Container.dispatchEventImpl(Container.java:2294)
	at java.awt.Component.dispatchEvent(Component.java:4713)
	at java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4888)
	at java.awt.LightweightDispatcher.processMouseEvent(Container.java:4525)
	at java.awt.LightweightDispatcher.dispatchEvent(Container.java:4466)
	at java.awt.Container.dispatchEventImpl(Container.java:2280)
	at java.awt.Window.dispatchEventImpl(Window.java:2750)
	at java.awt.Component.dispatchEvent(Component.java:4713)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:76)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.awt.EventQueue$4.run(EventQueue.java:731)
	at java.awt.EventQueue$4.run(EventQueue.java:729)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:76)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
````

### PR DESCRIPTION
**Pull request by stefan-kolb**
See #1455

### CODE REVIEW

File: src/main/java/net/sf/jabref/importer/fetcher/DOItoBibTeXFetcher.java [hunk 1]
```diff
@@ -86,14 +85,18 @@ public JPanel getOptionsPanel() {
...
+    public BibEntry getEntryFromDOI(String doiStr, ParserResult parserResult) {
         DOI doi;
         try {
             doi = new DOI(doiStr);
         } catch (IllegalArgumentException e) {
-            parserResult.addWarning(Localization.lang("Invalid DOI: '%0'.", doiStr));
+            if (parserResult != null) {
+                parserResult.addWarning(Localization.lang("Invalid DOI: '%0'.", doiStr));
+            }
             return null;
```
Comment (simonharrer) [inline src/main/java/net/sf/jabref/importer/fetcher/DOItoBibTeXFetcher.java]: Could this be replaced with Optional<Bibentry> as a return type? We try to minimize the usage of null.

**Comment by simonharrer** [comment]
LGTM

File: src/main/java/net/sf/jabref/importer/fetcher/DOItoBibTeXFetcher.java [hunk 2]
```diff
@@ -86,14 +85,18 @@ public JPanel getOptionsPanel() {
         return null;
     }
 
-    public BibEntry getEntryFromDOI(String doiStr, ParserResult parserResult) {
-        Objects.requireNonNull(parserResult);
+    public BibEntry getEntryFromDOI(String doiStr) {
+        return getEntryFromDOI(doiStr, null);
```
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/importer/fetcher/DOItoBibTeXFetcher.java]: I think the idea of the ParserResult variable is to communicate error messages. So users of this new overload method are not able to display meaningful error messages. Wouldn't it be better to change the corresponding places to something like:
````
ParserResult result = new ();
getEntryFromDOI(doi, result);
if(result.hasWarnings()) { displayThem() }
else { party() }
````

Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/importer/fetcher/DOItoBibTeXFetcher.java]: Some places just don't display error but solely use the fetcher. previously only null was passed for the ParserResult. I thought overloading the method makes more sense, but we can just delete the overloaded method again.

**Pull request merged**